### PR TITLE
add linux license check

### PR DESCRIPTION
**Pull request by gaohannk**

### CODE REVIEW

File: imagetest/test_suites/image_validation/linux_lincense_test.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+var licenseSpecPairs = map[string]string{
+	"centos-7":    "centos-7",
+	"centos-8":    "centos-8",
+	"rhel-7":      "rhel-7-server",
+	"rhel-8":      "rhel-8-server",
+	"rhel-7-byos": "rhel-7-byos",
+	"rhel-8-byos": "rhel-8-byos",
+	"debian-9":    "debian-9-stretch",
+	"debian-10":   "debian-10-stretch",
```
Comment (hopkiw) [inline imagetest/test_suites/image_validation/linux_lincense_test.go:22]: probably s/stretch/buster ?

File: imagetest/test_suites/image_validation/linux_lincense_test.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package imagevalidation
```
Comment (hopkiw) [inline imagetest/test_suites/image_validation/linux_lincense_test.go:1]: filename is misspelled

File: imagetest/test_suites/image_validation/linux_lincense_test.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"unicode"
+
+	"github.com/GoogleCloudPlatform/guest-test-infra/imagetest/utils"
+	"google.golang.org/api/compute/v1"
```
Comment (hopkiw) [inline imagetest/test_suites/image_validation/linux_lincense_test.go:11]: you can't use APIs from tests. the test VM will not be guaranteed to have permission to call this API.

**Review by hopkiw — CHANGES_REQUESTED** [review]

File: imagetest/test_suites/image_validation/linux_license_test.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	"strings"
+	"testing"
+	"unicode"
+
+	"github.com/GoogleCloudPlatform/guest-test-infra/imagetest/utils"
+)
+
+var imageLicenseCodeMap = map[string]string{
```
Comment (gaohannk) [inline imagetest/test_suites/image_validation/linux_license_test.go:11]: I haven't test all the license code. I check the license API and map them based on the license name

Comment (hopkiw) [reply imagetest/test_suites/image_validation/linux_license_test.go:11]: it should be easy to run this test against all these images and see if they all succeed

File: imagetest/test_suites/image_validation/linux_license_test.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		t.Fatalf("Failed to get license code metadata")
+	}
+	image, err := utils.GetMetadata("image")
+	if err != nil {
+		t.Fatalf("Failed to get image metadata")
+	}
+
+	// Example: projects/rhel-cloud/global/images/rhel-8-v20210217
+	splits := strings.Split(image, "/")
+	imageName := splits[4]
```
Comment (hopkiw) [inline imagetest/test_suites/image_validation/linux_license_test.go:60]: i think you should use `strings.SplitN` and explicitly check the length of the returned slice. if somehow this input was different, `imageName := splits[4]` could panic

i also think the method to pull out a base image name from the image metadata could be a utility in utils

File: imagetest/test_suites/image_validation/linux_license_test.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	imagePrefix := strings.TrimRightFunc(imageName, func(r rune) bool {
+		return unicode.IsDigit(r) || r == 'v' || r == '-'
+	})
+
+	if code, found := imageLicenseCodeMap[imagePrefix]; found == true && code == licenseCode {
+		t.Logf("Image %s has licenseCode %s", imageName, licenseCode)
+	}
+	t.Fatalf("Image has unkown licenseCode %s", licenseCode)
```
Comment (hopkiw) [inline imagetest/test_suites/image_validation/linux_license_test.go:68]: again i know in application programming this paradigm makes sense, but in tests success is the default and we only need to call out failures. and it isn't an 'unknown' licenseCode, it is an *incorrect* licenseCode. so i would flip this to check for failure as:
```
if code, found := codemap[image]; !found || code != expectedLicenseCode {
    t.Fatalf("Image %s has incorrect license code %d", imageName, expectedLicenseCode)
}
```

File: imagetest/test_suites/image_validation/linux_license_test.go (hunk 4 above)
Comment (gaohannk) [reply imagetest/test_suites/image_validation/linux_license_test.go:10]: <testsuite name="image_validation-ubuntu-minimal-2004-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-2004-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-1604-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-1804-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-minimal-1804-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7-4-sap" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7-6-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-minimal-1604-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7-7-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-8-2-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-8" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-centos-8" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-centos-7" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-debian-9" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-8-1-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.02"></testcase>
	</testsuite>
	<testsuite name="image_validation-debian-10" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
	<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
</testsuites>

**Comment by gaohannk** [comment]
<testsuites name="" errors="0" failures="0" tests="0" time="0">
	<testsuite name="image_validation-ubuntu-minimal-2004-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-2004-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-1604-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-1804-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-minimal-1804-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7-4-sap" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7-6-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-minimal-1604-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7-7-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-8-2-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
		<testsuite name="image_validation-debian-10" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-8" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-centos-8" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-centos-7" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-debian-9" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-8-1-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.02"></testcase>
	</testsuite>
</testsuites>

**Comment by gaohannk** [comment]
<testsuites name="" errors="0" failures="0" tests="0" time="0">
	<testsuite name="image_validation-ubuntu-minimal-2004-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-2004-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-1604-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-1804-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-minimal-1804-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7-4-sap" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7-6-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-ubuntu-minimal-1604-lts" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7-7-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-8-2-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-7" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.01"></testcase>
	</testsuite>
	<testsuite name="image_validation-debian-10" tests="1" failures="0" errors="1" disabled="0" skipped="0" time="0">
		<system-err>step &#34;create-disks&#34; validation error: cannot create disk &#34;vm1&#34;: can&#39;t use image &#34;projects/debian-cloud/global/images/family/debian-10&#34;: APIError: googleapi: Error 403: Quota exceeded for quota group &#39;ListGroup&#39; and limit &#39;List requests per 100 seconds&#39; of service &#39;compute.googleapis.com&#39; for consumer &#39;project_number:707281592825&#39;., rateLimitExceeded</system-err>
	</testsuite>
	<testsuite name="image_validation-rhel-8" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-centos-8" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-centos-7" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-debian-9" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
	</testsuite>
	<testsuite name="image_validation-rhel-8-1-sap-ha" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0.02"></testcase>
	</testsuite>
</testsuites>
	
<testsuite name="image_validation-debian-10" tests="1" failures="0" errors="0" disabled="0" skipped="0" time="0">
		<testcase classname="" name="TestLinuxLicense" time="0"></testcase>
</testsuite>

File: imagetest/test_suites/image_validation/linux_license_test.go (hunk 4 above)
Comment (gaohannk) [reply imagetest/test_suites/image_validation/linux_license_test.go:10]: Add test result in comment

File: imagetest/test_suites/image_validation/linux_license_test.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+	imageName, err := utils.ExtractBaseImageName(image)
+	if err != nil {
+		t.Fatal(err)
+	}
+	splits := strings.Split(imageName, "-")
+	if len(splits) < 2 {
+		t.Fatal("malformed base image name")
+	}
+	imagePrefix := strings.Join(splits[:len(splits)-1], "-")
```
Comment (hopkiw) [inline imagetest/test_suites/image_validation/linux_license_test.go:65]: i think this logic belongs in ExtractBaseImageName - given 'projects/rhel-cloud/global/images/rhel-8-v20210217' you would want to get back 'rhel-8'

**Pull request merged**